src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -0,0 +1,298 @@
+                var cancellationToken = _tokenSource.Token;
+                Task.Delay(TimeSpan.FromSeconds(1), cancellationToken)
+                    .ContinueWith(_ => OnWorkspaceChangedOnForeground(cancellationToken), cancellationToken, TaskContinuationOptions.OnlyOnRanToCompletion, this.ForegroundTaskScheduler);
+            }
+        }
+
+        private void OnWorkspaceChangedOnForeground(CancellationToken cancellationToken)
+        {
+            this.AssertIsForeground();
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return;
+            }
+
+            if (_workspace == null || _packageInstallerServices == null)
+            {
+                return;
+            }
+
+            var solution = _workspace.CurrentSolution;
+            ProjectId projectId;
+            bool clearAllData;
+            DequeueNextProject(solution, out projectId, out clearAllData);
+
+            if (clearAllData)
+            {
+                // Just clear out everything we have and start over in the case of a solution change.
+                _projectToInstalledPackageAndVersion.Clear();
+            }
+
+            if (projectId == null)
+            {
+                // No project to process, nothing to do.
+                return;
+            }
+
+            // Process this single project.
+            ProcessProjectChange(solution, projectId);
+
+            // After processing this single project, yield so the foreground thread
+            // can do more work.  Then go and loop again so we can process the 
+            // rest of the projects.
+            Task.Factory.SafeStartNew(
+                () => OnWorkspaceChangedOnForeground(cancellationToken), cancellationToken, ForegroundTaskScheduler);
+        }
+
+        private void DequeueNextProject(Solution solution, out ProjectId projectId, out bool clearAllData)
+        {
+            this.AssertIsForeground();
+
+            lock (_gate)
+            {
+                // If we detected a solution change, then add all projects to the list of projects
+                // to process.
+                clearAllData = _solutionChanged;
+                if (_solutionChanged)
+                {
+                    _changedProjects.AddRange(solution.ProjectIds);
+                }
+
+                _solutionChanged = false;
+
+                // Remove and return the first project in the list.
+                projectId = _changedProjects.FirstOrDefault();
+                _changedProjects.Remove(projectId);
+            }
+        }
+
+        private void ProcessProjectChange(Solution solution, ProjectId projectId)
+        {
+            this.AssertIsForeground();
+
+            // Remove anything we have associated with this project.
+            Dictionary<string, string> installedPackages;
+            _projectToInstalledPackageAndVersion.TryRemove(projectId, out installedPackages);
```

Comment:
Can't you just do `.Remove(projectId)`? and not have the used `installedPackages` return value?